File: src/server/manager.ts
import { StateVector, propagateRK4, magnitude, subtract, normalize, add, multiply } from "../lib/physics.ts";
// @ts-ignore
import kdTree from "kd-tree-javascript";

export interface SpaceObject {
  id: string;
  type: "satellite" | "debris";
  state: StateVector;
  mass: number; // kg
  fuel: number; // kg
  maxFuel: number;
  assignedOrbit?: StateVector;
  maneuvers: Maneuver[];
  lastManeuverTime: number;
}

export interface Maneuver {
  time: number;
  deltaV: { x: number; y: number; z: number };
  type: "evasion" | "recovery" | "graveyard";
}

export interface CollisionRisk {
  obj1Id: string;
  obj2Id: string;
  distance: number;
  timeToCollision: number;
}

export class ACMManager {
  private objects: Map<string, SpaceObject> = new Map();
  private currentTime: number = 0;
  private groundStations = [
    { name: "Svalbard", lat: 78.2, lon: 15.6 },
    { name: "McMurdo", lat: -77.8, lon: 166.6 },
    { name: "Troll", lat: -72.0, lon: 2.5 }
  ];

  constructor() {
    this.seedInitialData();
  }

  private seedInitialData() {
    // Add some satellites
    for (let i = 0; i < 50; i++) {
      const alt = 500000 + Math.random() * 100000;
      const r = 6378137 + alt;
      const v = Math.sqrt(3.986004418e14 / r);
      const angle = (i / 50) * Math.PI * 2;
      
      const state: StateVector = {
        pos: { x: r * Math.cos(angle), y: r * Math.sin(angle), z: 0 },
        vel: { x: -v * Math.sin(angle), y: v * Math.cos(angle), z: 0 }
      };

      this.objects.set(`SAT-${i}`, {
        id: `SAT-${i}`,
        type: "satellite",
        state: { ...state },
        mass: 500,
        fuel: 100,
        maxFuel: 100,
        assignedOrbit: { ...state },
        maneuvers: [],
        lastManeuverTime: -1000
      });
    }

    // Add debris
    for (let i = 0; i < 2000; i++) { // Starting with 2000 for stability
      const alt = 400000 + Math.random() * 300000;
      const r = 6378137 + alt;
      const v = Math.sqrt(3.986004418e14 / r);
      const angle = Math.random() * Math.PI * 2;
      const inc = (Math.random() - 0.5) * Math.PI;

      this.objects.set(`DEB-${i}`, {
        id: `DEB-${i}`,
        type: "debris",
        state: {
          pos: { x: r * Math.cos(angle), y: r * Math.sin(angle) * Math.cos(inc), z: r * Math.sin(angle) * Math.sin(inc) },
          vel: { x: -v * Math.sin(angle), y: v * Math.cos(angle) * Math.cos(inc), z: v * Math.cos(angle) * Math.sin(inc) }
        },
        mass: 1,
        fuel: 0,
        maxFuel: 0,
        maneuvers: [],
        lastManeuverTime: 0
      });
    }
  }

  ingestTelemetry(objects: SpaceObject[]) {
    objects.forEach(obj => this.objects.set(obj.id, obj));
  }

  scheduleManeuver(satelliteId: string, maneuver: Maneuver) {
    const sat = this.objects.get(satelliteId);
    if (sat && sat.type === "satellite") {
      sat.maneuvers.push(maneuver);
    }
  }

  step(dt: number) {
    this.currentTime += dt;
    const objectsArray = Array.from(this.objects.values());

    // 1. Propagate all objects
    objectsArray.forEach(obj => {
      // Check for scheduled maneuvers
      const activeManeuver = obj.maneuvers.find(m => Math.abs(m.time - this.currentTime) < dt / 2);
      if (activeManeuver && obj.fuel > 0) {
        // Apply deltaV
        obj.state.vel = add(obj.state.vel, activeManeuver.deltaV);
        // Rocket equation: deltaV = ve * ln(m0/m1) => m1 = m0 / exp(deltaV/ve)
        // Simplified fuel consumption: 1kg per 10m/s for 500kg sat
        const dvMag = magnitude(activeManeuver.deltaV);
        const fuelConsumed = (dvMag / 10) * (obj.mass / 500);
        obj.fuel = Math.max(0, obj.fuel - fuelConsumed);
        obj.lastManeuverTime = this.currentTime;
        obj.maneuvers = obj.maneuvers.filter(m => m !== activeManeuver);
      }

      obj.state = propagateRK4(obj.state, dt);
    });

    // 2. Collision Detection using KD-Tree
    const treeData = objectsArray.map(obj => ({
      id: obj.id,
      x: obj.state.pos.x,
      y: obj.state.pos.y,
      z: obj.state.pos.z,
      obj: obj
    }));

    const tree = new kdTree.kdTree(treeData, (a: any, b: any) => {
      const dx = a.x - b.x;
      const dy = a.y - b.y;
      const dz = a.z - b.z;
      return dx * dx + dy * dy + dz * dz;
    }, ["x", "y", "z"]);

    const risks: CollisionRisk[] = [];
    objectsArray.filter(o => o.type === "satellite").forEach(sat => {
      const satNode = { x: sat.state.pos.x, y: sat.state.pos.y, z: sat.state.pos.z };
      const nearest = tree.nearest(satNode, 2, 1000 * 1000); // 1km radius
      nearest.forEach(([node, distSq]: [any, number]) => {
        if (node.id !== sat.id) {
          const dist = Math.sqrt(distSq);
          if (dist < 100) {
            risks.push({ obj1Id: sat.id, obj2Id: node.id, distance: dist, timeToCollision: 0 });
          }
        }
      });
    });

    // 3. Autonomous Logic
    this.runAutonomousLogic(risks);

    return {
      currentTime: this.currentTime,
      risks,
      stats: {
        satellites: objectsArray.filter(o => o.type === "satellite").length,
        debris: objectsArray.filter(o => o.type === "debris").length
      }
    };
  }

  private runAutonomousLogic(risks: CollisionRisk[]) {
    risks.forEach(risk => {
      const sat = this.objects.get(risk.obj1Id);
      if (sat && sat.type === "satellite" && this.currentTime - sat.lastManeuverTime > 600) {
        // Calculate evasion
        // Prefer transverse direction (perpendicular to velocity and position)
        const radial = normalize(sat.state.pos);
        const velocity = normalize(sat.state.vel);
        const transverse = {
          x: radial.y * velocity.z - radial.z * velocity.y,
          y: radial.z * velocity.x - radial.x * velocity.z,
          z: radial.x * velocity.y - radial.y * velocity.x
        };

        const deltaV = multiply(transverse, 5); // 5 m/s evasion
        
        // Check fuel and graveyard
        if (sat.fuel / sat.maxFuel < 0.05) {
          this.scheduleManeuver(sat.id, {
            time: this.currentTime + 10,
            deltaV: multiply(velocity, 15), // Boost to graveyard
            type: "graveyard"
          });
        } else {
          this.scheduleManeuver(sat.id, {
            time: this.currentTime + 10,
            deltaV,
            type: "evasion"
          });
          // Schedule recovery
          this.scheduleManeuver(sat.id, {
            time: this.currentTime + 1200,
            deltaV: multiply(deltaV, -1),
            type: "recovery"
          });
        }
      }
    });

    // Station keeping
    this.objects.forEach(obj => {
      if (obj.type === "satellite" && obj.assignedOrbit && this.currentTime - obj.lastManeuverTime > 3600) {
        const drift = magnitude(subtract(obj.state.pos, obj.assignedOrbit.pos));
        if (drift > 10000) { // 10km drift
          const correction = multiply(normalize(subtract(obj.assignedOrbit.pos, obj.state.pos)), 2);
          this.scheduleManeuver(obj.id, {
            time: this.currentTime + 60,
            deltaV: correction,
            type: "recovery"
          });
        }
      }
    });
  }

  getSnapshot() {
    return {
      time: this.currentTime,
      objects: Array.from(this.objects.values()).map(o => ({
        id: o.id,
        type: o.type,
        pos: o.state?.pos || { x: 0, y: 0, z: 0 },
        fuel: o.fuel,
        maxFuel: o.maxFuel
      }))
    };
  }
}

File: manager.py
import numpy as np
from scipy.spatial import KDTree
from physics import propagate_rk4
import time

class ACMManager:
    def __init__(self):
        self.objects = {}
        self.current_time = 0
        self.ground_stations = [
            {"name": "Svalbard", "lat": 78.2, "lon": 15.6},
            {"name": "McMurdo", "lat": -77.8, "lon": 166.6},
            {"name": "Troll", "lat": -72.0, "lon": 2.5}
        ]
        self.seed_initial_data()

    def has_los(self, sat_pos):
        # Calculate satellite lat/lon
        r = np.linalg.norm(sat_pos)
        lat = np.degrees(np.arcsin(sat_pos[2] / r))
        lon = np.degrees(np.arctan2(sat_pos[1], sat_pos[0]))
        
        # Simple LOS check: Is the satellite above the horizon for any station?
        for station in self.ground_stations:
            d_lat = abs(lat - station["lat"])
            d_lon = abs(lon - station["lon"])
            if d_lat < 20 and d_lon < 20:
                # Blackout Zone check (e.g., South Atlantic Anomaly simulation)
                if -50 < lon < 0 and -40 < lat < 0:
                    return False # Blackout zone
                return True
        return False

    def seed_initial_data(self):
        # Add satellites
        for i in range(50):
            alt = 500000 + np.random.random() * 100000
            r = 6378137 + alt
            v = np.sqrt(3.986004418e14 / r)
            angle = (i / 50.0) * np.pi * 2
            
            pos = np.array([r * np.cos(angle), r * np.sin(angle), 0.0])
            vel = np.array([-v * np.sin(angle), v * np.cos(angle), 0.0])

            self.objects[f"SAT-{i}"] = {
                "id": f"SAT-{i}",
                "type": "satellite",
                "pos": pos,
                "vel": vel,
                "mass": 500.0,
                "fuel": 100.0,
                "max_fuel": 100.0,
                "assigned_orbit": {"pos": pos.copy(), "vel": vel.copy()},
                "maneuvers": [],
                "last_maneuver_time": -1000
            }

        # Add debris
        for i in range(2000):
            alt = 400000 + np.random.random() * 300000
            r = 6378137 + alt
            v = np.sqrt(3.986004418e14 / r)
            angle = np.random.random() * np.pi * 2
            inc = (np.random.random() - 0.5) * np.pi

            pos = np.array([r * np.cos(angle), r * np.sin(angle) * np.cos(inc), r * np.sin(angle) * np.sin(inc)])
            vel = np.array([-v * np.sin(angle), v * np.cos(angle) * np.cos(inc), v * np.cos(angle) * np.sin(inc)])

            self.objects[f"DEB-{i}"] = {
                "id": f"DEB-{i}",
                "type": "debris",
                "pos": pos,
                "vel": vel,
                "mass": 1.0,
                "fuel": 0.0,
                "max_fuel": 0.0,
                "maneuvers": [],
                "last_maneuver_time": 0
            }

    def ingest_telemetry(self, telemetry_data):
        for obj in telemetry_data:
            oid = obj["id"]
            self.objects[oid] = {
                "id": oid,
                "type": obj["type"],
                "pos": np.array([obj["x"], obj["y"], obj["z"]]),
                "vel": np.array([obj["vx"], obj["vy"], obj["vz"]]),
                "mass": obj.get("mass", 1.0),
                "fuel": obj.get("fuel", 0.0),
                "max_fuel": obj.get("max_fuel", 0.0),
                "maneuvers": [],
                "last_maneuver_time": 0
            }

    def schedule_maneuver(self, satellite_id, burn):
        if satellite_id in self.objects:
            self.objects[satellite_id]["maneuvers"].append(burn)

    def step(self, dt):
        self.current_time += dt
        obj_list = list(self.objects.values())

        # 1. Propagate
        for obj in obj_list:
            # Apply maneuvers
            remaining_maneuvers = []
            for m in obj["maneuvers"]:
                if abs(m["time"] - self.current_time) < dt / 2:
                    dv = np.array([m["dv_x"], m["dv_y"], m["dv_z"]])
                    dv_mag = np.linalg.norm(dv)
                    if obj["fuel"] > 0:
                        obj["vel"] += dv
                        fuel_consumed = (dv_mag / 10.0) * (obj["mass"] / 500.0)
                        obj["fuel"] = max(0.0, obj["fuel"] - fuel_consumed)
                        obj["last_maneuver_time"] = self.current_time
                else:
                    remaining_maneuvers.append(m)
            obj["maneuvers"] = remaining_maneuvers

            obj["pos"], obj["vel"] = propagate_rk4(obj["pos"], obj["vel"], dt)

        # 2. Collision Detection
        positions = np.array([obj["pos"] for obj in obj_list])
        tree = KDTree(positions)
        
        risks = []
        for i, obj in enumerate(obj_list):
            if obj["type"] == "satellite":
                indices = tree.query_ball_point(obj["pos"], 1000.0) # 1km
                for idx in indices:
                    if idx != i:
                        other = obj_list[idx]
                        dist = np.linalg.norm(obj["pos"] - other["pos"])
                        if dist < 100:
                            risks.append({
                                "obj1Id": obj["id"],
                                "obj2Id": other["id"],
                                "distance": float(dist),
                                "timeToCollision": 0
                            })

        # 3. Autonomous Logic
        self.run_autonomous_logic(risks)

        return {
            "currentTime": self.current_time,
            "risks": risks,
            "stats": {
                "satellites": len([o for o in obj_list if o["type"] == "satellite"]),
                "debris": len([o for o in obj_list if o["type"] == "debris"])
            }
        }

    def run_autonomous_logic(self, risks):
        for risk in risks:
            sat = self.objects.get(risk["obj1Id"])
            if sat and sat["type"] == "satellite" and (self.current_time - sat["last_maneuver_time"]) > 600:
                # Check Communication LOS
                if not self.has_los(sat["pos"]):
                    continue # Cannot maneuver without ground link
                
                # Evasion
                pos_norm = sat["pos"] / np.linalg.norm(sat["pos"])
                vel_norm = sat["vel"] / np.linalg.norm(sat["vel"])
                transverse = np.cross(pos_norm, vel_norm)
                
                dv = transverse * 5.0
                
                if sat["fuel"] / sat["max_fuel"] < 0.05:
                    # Graveyard
                    self.schedule_maneuver(sat["id"], {
                        "time": self.current_time + 10,
                        "dv_x": float(vel_norm[0] * 15),
                        "dv_y": float(vel_norm[1] * 15),
                        "dv_z": float(vel_norm[2] * 15),
                        "type": "graveyard"
                    })
                else:
                    self.schedule_maneuver(sat["id"], {
                        "time": self.current_time + 10,
                        "dv_x": float(dv[0]), "dv_y": float(dv[1]), "dv_z": float(dv[2]),
                        "type": "evasion"
                    })
                    # Recovery
                    self.schedule_maneuver(sat["id"], {
                        "time": self.current_time + 1200,
                        "dv_x": float(-dv[0]), "dv_y": float(-dv[1]), "dv_z": float(-dv[2]),
                        "type": "recovery"
                    })

    def get_snapshot(self):
        return {
            "time": self.current_time,
            "objects": [
                {
                    "id": o["id"],
                    "type": o["type"],
                    "pos": {"x": float(o["pos"][0]), "y": float(o["pos"][1]), "z": float(o["pos"][2])},
                    "fuel": float(o["fuel"]),
                    "maxFuel": float(o["max_fuel"])
                } for o in self.objects.values()
            ]
        }
